### ISSUE TITLE
Some Downloads not work anymore

### ISSUE THREAD

**Issue by Dragodraki** [wingman-jr-addon/wingman_jr#202]
On rare pages downloads not work anymore with Wingman Jr. (my current is 3.3.6) enabled (Firefox v. 115.8.0 esr).
E.g. here: https://www.ibiblio.org/pub/micro/pc-stuff/freedos/files/distributions/1.0/

When clicking on a link, instead of a download the website shows cryptic content [STRG+S would be NOT the download]. This would have to be fixed urgently. Can you do this?

**Comment by wingman-jr-addon** [comment]
Thanks for the report @Dragodraki @SufianBabri .
@Dragodraki  I think I can reproduce what you're talking about for the ISO's there - for example https://www.ibiblio.org/pub/micro/pc-stuff/freedos/files/distributions/1.0/fdboot.img
@SufianBabri I'm getting the following XML even without Wingman Jr.:
`<Error>
<Code>AccessDenied</Code>
<Message>Access denied.</Message>
</Error>`
I'm assuming that's not what you'd expect?

At any rate, I'll take a peek and see if I can figure out what's going on.

**Comment by wingman-jr-addon** [comment]
Working on bisecting:

- 3.3.6 - As Dragodraki described, likely just not triggering download
- 3.3.0 - Same as 3.3.6 but not correctly replacing some of the characters
- 3.0.0 - Same as 3.3.0
- 2.0.1 - Same as 3.0.0

So, at least this doesn't seem to be a recent regression. My guess is its something to do with the content type handling for e.g. octet-stream and friends but we'll find out.

**Comment by wingman-jr-addon** [comment]
Ok, so the core issue seems to be that the .IMG isn't serving up a Content-Type at all. Here's what Wingman does:
![image](https://github.com/wingman-jr-addon/wingman_jr/assets/55339824/471fb322-29a7-43e7-84ca-df13ad8b0484)

I found a similar site, with similar type of content: http://ftp.vcu.edu/pub/gnu_linux/archlinux/iso/2024.04.01/
Here's what it does instead:
![image](https://github.com/wingman-jr-addon/wingman_jr/assets/55339824/d1e52054-3b10-4896-bc71-ed7b302219f3)

The `application/octet-stream` will trigger it to download.

So, big picture: Wingman Jr. has to do things around Content-Type so that it can properly translate/pass through characters. To do this, it has to force a specific content type. However, in this case, the original Content-Type is not specified and so then Firefox presumably runs its own smarter content type detection and determines it should download.

Now that might be the problem, but I'm not sure what the solution should be yet.

**Comment by wingman-jr-addon** [comment]
As a further resource for later, the "what to do when type isn't specified" can get _complex_, see for example https://mimesniff.spec.whatwg.org/#identifying-a-resource-with-an-unknown-mime-type

**Comment by Dragodraki** [comment]
> Thanks for the report @Dragodraki @SufianBabri . @Dragodraki I think I can reproduce what you're talking about for the ISO's there - for example https://www.ibiblio.org/pub/micro/pc-stuff/freedos/files/distributions/1.0/fdboot.img @SufianBabri I'm getting the following XML even without Wingman Jr.: `<Error> <Code>AccessDenied</Code> <Message>Access denied.</Message> </Error>` I'm assuming that's not what you'd expect?
> 
> At any rate, I'll take a peek and see if I can figure out what's going on.

If the website does not work for you, maybe is is about regional limitation (geo-lock) - with VPN or TOR you should bypass that. Sorry I cannot give another example right now.

With Wingman Jr enabled the website shows trys to display download content as plain text instead of offer download - here is an excerpt of the weird characters:

"ë<LINUX4.1à@ð )ãúD FAT12 úü1ÀŽØ½|¸àŽÀ‰î‰ï¹ó¥ê^|à`ŽØŽÐf û€~$ÿuˆV$ÇFÀÇFÂèéFreeDOS‹v‹~vƒ×‰vÒ‰~ÔŠF˜÷fÆ×‰vÖ‰~Ø‹^±Óë‹F1Ò÷ó‰FÐÆƒ×‰vÚ‰~Ü‹FÖ‹VØ‹~ÐÄ^Zè›r/Ä~Z¹¾ñ}Wó¦_&‹EtƒÇ &€=uçrYPÄ^Z‹~‹FÒ‹VÔèk"

Normally it has to look like this (when disable the addon it does):
<img width="580" alt="Screenshot 2024-04-07 140359" src="https://github.com/wingman-jr-addon/wingman_jr/assets/76787321/ce9f893a-f3a8-45b7-ba65-ad587353d40b">

**Comment by wingman-jr-addon** [comment]
Yep, thanks @Dragodraki that is what I see too for your example. As noted above, the root of the issue is that the website doesn't send a Content-Type. Wingman Jr adds one, but then that means that Firefox can't use its own logic to properly infer a Content-Type. Usually downloads have a type of "application/octet-stream", but in this case Wingman Jr is incorrectly inferring a text type. I'm still trying to think the best way to actually fix this.

**Comment by wingman-jr-addon** [comment]
So mulling on this a bit further, apart from thinking how to modify the code the currently exists, there is still a core conundrum. For documents that don't supply a Content-Type, the default behaviour of browsers is to treat some as documents and some as downloads (as noted previously). Ones that are treated as documents should definitely be scanned, but ones that are not should probably go through the normal download process. However, in order to know which are which, Wingman cannot just defer the guessing process to Firefox. So, the only viable path through is to actually implement Content-Type sniffing. However, I do wonder how thorny this will get: while the stated standard algorithm is complex, I'm concerned that browsers may add their own extra detection logic on top and that forever I'll be reverse engineering that logic; still, it's probably better than the current state.

With respect to implementation, then, there's definitely work to do. Right now only the charset is sniffed, and that is in part dependent on Content-Type detection happening as a prerequisite. Now the case where Content-Type itself is sniffed has to be handled, and early return on the logic based on Content-Type appearing in the header won't suffice, so it could get yucky. However, at least it's clear from the above that Content-Type sniffing _must_ be reimplemented for the rest to work correctly.

Both this and #201 are similar in that the API to do the scanning doesn't really expose enough of what the browser is doing and both require essentially re-implementing a core and complex part of what the browser itself does.

**Comment by arthurmelton** [comment]
I don't know if this would be the best way of implementing this, but you can have a look at the [`file` command source code](https://github.com/file/file/tree/master/magic/Magdir). It has magic definitions for pretty much every type of thing imaginable. It uses a "formatting" language to read some specific configuration files to use for checking the type of input. They have support for getting the name, mime, and ext of the data. I doubt you would want to try and write your own interpreter, but with the name being so generic I could not find any libraries to use this in JavaScript.

**Comment by Dragodraki** [comment]
First of all, thanks to everyone for helping in doing comments and providing details related to this issue.

@wingman-jr-addon 
The addon don't have to be perfect. Like for so many things in life, perfection should be the aim but not the reality. Maybe you can make Wingman Jr. a little bit smarter interpreting the content type but only as long the results are worth it.
If you find a solution for the more common content types - like in my example - I will be fully satfisfied (you might close the issue then).

**Comment by wingman-jr-addon** [comment]
Thanks @Dragodraki @arthurmelton . I did finally fix #201 , and I think this has some similarities to it, with a key difference being that I can't "sniff" the MIME type after passing on the start of the request, I have to "sniff" before - but preferably only in the case when no MIME type is specified. So, I can probably implement just 7.1 from here: https://mimesniff.spec.whatwg.org/#mime-type-sniffing-algorithm

**Comment by Dragodraki** [comment]
Here is a list of webpages/downloads that wont be displayed with having Wingman Jr enabled (v.3.4.0):
This list is not the newest list, please take a look at later posts.

- https://www.os2museum.com/files/videomp4.img
- https://cdn.driverscape.com/2015/display/displaylink_usb_7_7m4.zip?access_key=1726428315-q%2BvBQZYfgq2Burm9HC05Lid%2FX2VPbPDJDrgxPjhrhgs%3D

It shows the special character thing again here or loads indefinitely.

**Comment by wingman-jr-addon** [comment]
Thanks @Dragodraki , I still have not tried to tackle this but having known failures to try makes development easier so thanks for reporting.

**Comment by Dragodraki** [comment]
Everytime, and thanks for your continous support in this addon. It's great improvement over the last years is outstanding! Take your time.

**Comment by wingman-jr-addon** [comment]
Ok, so I'm continuing to noodle on this. Notes for myself regarding architecture:

- bkDirectTypedUrlListener and bkBase64ContentListener should probably merge
- General flow will be to check the following in order:

1. If Content-Type present and image, fall back to normal image listener
2. If Content-Type present and some HTML type, do base-64 listener.
3. If Content-Type present but none of the above, don't filter.
4. Otherwise sniff for MIME type per 7.1 as noted previously.
5. If sniffed MIME type is present and image, tap into normal image listener. This could get a little tricky since we've already received data, but could maybe simulate by sending the first packet received in.
6. If sniffed MIME type is present and HTML type, tap into normal base-64 listener. Similar to above, may have to simulate normal flow by forwarding on the first packet.
7. If sniffed MIME type is other type or unknown, treat as download.

**Comment by Dragodraki** [comment]
Another website where this error occurs:
https://tihiy.net/files/OldNewExplorer.rar

**Comment by wingman-jr-addon** [comment]
(As a note, fix https://github.com/wingman-jr-addon/wingman_jr/issues/245 does not help this one)

**Comment by Dragodraki** [comment]
Here is a list of webpages/downloads that wont be displayed with having Wingman Jr enabled (v.3.8.0):
Last updated: 2nd August 2026

- https://www.os2museum.com/files/videomp4.img
- https://cdn.driverscape.com/2015/display/displaylink_usb_7_7m4.zip?access_key=1726428315-q%2BvBQZYfgq2Burm9HC05Lid%2FX2VPbPDJDrgxPjhrhgs%3D
- https://tihiy.net/files/OldNewExplorer.rar
- https://apps.steamcore.se/Screamer.win-x64/Releases/ScreamerSetup.exe

It shows the special character thing again here or loads indefinitely.

Maybe you could try to find a fix?

**Comment by wingman-jr-addon** [comment]
@Dragodraki While the full fix is still quite extensive as listed above, I thought about it and adding a "force download" right click menu item for links is a good step. Note that it will only show up if you have the on/off switch enabled, because allowing to force download acts as a bit of a bypass. You can see a draft here: https://github.com/wingman-jr-addon/wingman_jr/pull/252

**Comment by Dragodraki** [comment]
Thanks for your effort. Though I still think it might be the best to try finding an automatic solution. Having users too often using this any form of exception like the bypass via right click menu would be disturbing and could lead them to turn off Wingman Jr.

<img width="887" height="411" alt="Image" src="https://github.com/user-attachments/assets/c4bf006c-e7d6-456b-b234-df4726928680" />

<img width="1689" height="99" alt="Image" src="https://github.com/user-attachments/assets/e1c9a3e4-07e5-401d-a467-2544995bcc9a" />

Another reason for right click menu is not thre best is, that said problematic download interfere with Firefox in a way my (high-end) CPU get to its limits.
https://apps.steamcore.se/Screamer.win-x64/Releases/ScreamerSetup.exe
I don't know why that is so, but I assume two features fight against each other then. It only ends, if I close the tab.

Here is my suggestion:
Though it might be profane, you could just determine the URL - if it ends with typical media filetype endings like .png, .gif, .flac, .mp3, .mp4, .zip, .iso ... or the base64-encoded version of it, its high probably a download (even if its displayed in-browser) and Wingman Jr. should either turn off it's detection for this site if it's not an image or video. If the download is an image or video, it least the error message could be better than the picture above. There might be better ways to determine downloads, e.g. with inspector tool, but this solution would be simple AND relatively efficient.
How does it sound?

**Comment by wingman-jr-addon** [comment]
Hmmm, I think there could be something to that even though it's the "wrong" way to do it for sure.
One reason I haven't tried approaching it that way before is because there are mistakes that can happen the other way too. For example, maybe the extension is ".jpg" but the browser is actually being instructed to download it via e.g. the MIME type. So as a "positive" test it probably shouldn't be used.
But, I think it could work well as a "negative" test for known extensions. It would be quite rare indeed for a ".exe" to actually have a MIME type of an image. Could it happen? Yes. But would the addon make fewer mistakes I think by taking that approach.
Does that make sense?

**Comment by Dragodraki** [comment]
Yep, you nailed it. I think this way too. 
Maybe you could just pre-check, whether its file header matches with one of special binary files like EXE/DLL/OCX/ISO/ZIP/RAR/7Z/... which could not be scanned anyway and shouldn't. This would be resistent to false positives AND false negatives.
If that is absolutely not possible, I can really understand there is a limit to what you can inspect with addons, then lets do it via plain file type name.

**Comment by wingman-jr-addon** [comment]
@Dragodraki I have a draft up here: https://github.com/wingman-jr-addon/wingman_jr/pull/253

**Comment by Dragodraki** [comment]
Thank you, can you test it with the given websites? I doubt, Firefox/Chrome allow me to modify existing extensions without valid signature anymore like back in times of XUL-based addons.

**Comment by wingman-jr-addon** [comment]
@Dragodraki Yes, I was testing it with the website links above and it was working prior to me posting: ISO's, .exe's etc seemed to be downloading fine and I haven't noticed any side effects.

If you'd like me to test any other specific sites, just post them here.

Otherwise if you want to test yourself, the way I do it is to temporarily disable the current addon and then  a) clone the repository using git, b) switch to the branch in the pull request (in this case "ignore-binaries"), then c) go to about:debugging and pick the manifest.json from the checked out repository. Unless you have other corporate policies enabled, these newer addons don't require compiled / signed code if you test it this way. But, I understand if that's too much ... just telling you how I do it.

So far it's looking good, so if I don't hear back in a while I'll be planning to merge it.

**Comment by Dragodraki** [comment]
@wingman-jr-addon 
Thanks to your step-by-step instructions it was very easy for me to test it. Yes, it works with this branch - you can merge it. I can see you chose the easy way to bypass those files by a list of whitelisted file extensions.

Maybe, just maybe, you can get a function to check the file header instead of file extensions. Technically spoken it highly depends on whether a browser let you read the content of files in general before download prompt - the test itself should work: I instead of a whitelist of hardcoded filetypes it would be hardcoded headers hex sequences. This way, it would nearly cover all files (then there won't the risk of false positive or false negatives and it would cover far more scenarios), like the ancient windows software "FiletypeID"...

**Comment by Dragodraki** [comment]
When do you expect the new version to be released? Probably you collect several bugfixes from other issues for an update, but maybe you can give me an ETA?

**Comment by wingman-jr-addon** [comment]
@Dragodraki It's released! https://github.com/wingman-jr-addon/wingman_jr/releases/tag/release-3.9.0

**Comment by Dragodraki** [comment]
Thanks you very much - I already received th update :)
